### PR TITLE
HOSTEDCP-1121: Ensure SG reconciliation for aws endpoint

### PR DESCRIPTION
**Pull request by enxebre**
**What this PR does / why we need it**:
This ensures that clusters running before we enforced the hypershift security group on the endpoint creation get that SG as well. 

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

**Comment by enxebre** [comment]
cc @mjlshen

File: control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go [hunk 1]
```diff
@@ -446,14 +446,24 @@ func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointService(ctx context.C
...
+
+		// Ensure endpoint has the right SG.
+		exitingSG := make([]*string, 0)
+		for _, group := range output.VpcEndpoints[0].Groups {
+			exitingSG = append(exitingSG, group.GroupId)
+		}
+		addedSG, removedSG := diffIDs([]string{hcp.Status.Platform.AWS.DefaultWorkerSecurityGroupID}, exitingSG)
+
+		if addedSubnet != nil || removedSubnet != nil || addedSG != nil || removedSG != nil {
 			log.Info("endpoint subnets have changed")
```
Comment (mjlshen) [inline control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go:460]:
```suggestion
			log.Info("endpoint subnets or security groups have changed")
```

**Review by mjlshen — APPROVED** [review]
Thanks for taking care of this!

**Comment by enxebre** [comment]
HCs are failing to delete because the [CPO can't delete the SG](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2872/pull-ci-openshift-hypershift-main-e2e-aws/1687143829058097152/artifacts/e2e-aws/run-e2e/artifacts/TestCreateCluster_DestroyCluster_1/namespaces/e2e-clusters-8c99h-example-twnql/core/pods/logs/control-plane-operator-8cb4d458c-lvrx9-control-plane-operator.log)

```
{"level":"error","ts":"2023-08-03T17:53:12Z","msg":"Reconciler error","controller":"hostedcontrolplane","controllerGroup":"hypershift.openshift.io","controllerKind":"HostedControlPlane","HostedControlPlane":{"name":"example-twnql","namespace":"e2e-clusters-8c99h-example-twnql"},"namespace":"e2e-clusters-8c99h-example-twnql","name":"example-twnql","reconcileID":"1a8f55f8-267a-421c-b7ec-f21349a8e797","error":"failed to delete AWS default security group: failed to delete security group sg-0e617478f69845895: DependencyViolation: resource sg-0e617478f69845895 has a dependent object\n\tstatus code: 400, request id: 6d530e23-db27-45a7-805a-9cdbce8860fe","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/hypershift/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:326\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/hypershift/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/hypershift/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234"}
```

It can't delete the SG because the awsendpointservice is still using it, as [it is not being deleted](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2872/pull-ci-openshift-hypershift-main-e2e-aws/1687143829058097152/artifacts/e2e-aws/run-e2e/artifacts/TestCreateCluster_DestroyCluster_1/namespaces/e2e-clusters-8c99h-example-twnql/hypershift.openshift.io/awsendpointservices/private-router.yaml)

The awsendpointservice is not being deleted because the [capi Cluster resource hasn't finished deletion yet](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2872/pull-ci-openshift-hypershift-main-e2e-aws/1687143829058097152/artifacts/e2e-aws/run-e2e/artifacts/TestCreateCluster_DestroyCluster_1/namespaces/hypershift/core/pods/logs/operator-84b7774bbf-dfmgr-operator.log)
```
{"level":"info","ts":"2023-08-03T17:55:13Z","msg":"Waiting for cluster deletion","controller":"hostedcluster","controllerGroup":"hypershift.openshift.io","controllerKind":"HostedCluster","HostedCluster":{"name":"example-n679l","namespace":"e2e-clusters-jxt9t"},"namespace":"e2e-clusters-jxt9t","name":"example-n679l","reconcileID":"343d2ffb-83d1-43ef-a057-f5d1d4d3e81d","clusterName":"example-n679l","controlPlaneNamespace":"e2e-clusters-jxt9t-example-n679l"}
```

The capi Cluster resource hasn't finished deletion yet because [there's a leaked awsMachine](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2872/pull-ci-openshift-hypershift-main-e2e-aws/1687143829058097152/artifacts/e2e-aws/run-e2e/artifacts/TestCreateCluster_DestroyCluster_1/namespaces/e2e-clusters-8c99h-example-twnql/infrastructure.cluster.x-k8s.io/awsclusters/)

Why there's a leaked awsMachine?

**Comment by enxebre** [comment]
See if this makes a difference https://github.com/openshift/hypershift/pull/2872/files#diff-b5cde720842b19b53da3f454d827e27327c4559b2e4849906dac728586f51eb3R3272-R3274

File: control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go [hunk 2]
```diff
@@ -446,14 +446,24 @@ func (r *AWSEndpointServiceReconciler) reconcileAWSEndpointService(ctx context.C
...
+
+		if addedSubnet != nil || removedSubnet != nil || addedSG != nil || removedSG != nil {
+			log.Info("endpoint subnets or security groups have changed")
 			_, err := ec2Client.ModifyVpcEndpointWithContext(ctx, &ec2.ModifyVpcEndpointInput{
-				VpcEndpointId:   aws.String(endpointID),
-				AddSubnetIds:    added,
-				RemoveSubnetIds: removed,
+				VpcEndpointId:          aws.String(endpointID),
+				AddSubnetIds:           addedSubnet,
+				RemoveSubnetIds:        removedSG,
```
Comment (enxebre) [inline control-plane-operator/controllers/awsprivatelink/awsprivatelink_controller.go:464]: `RemoveSubnetIds: removedSG,`
bad ^^. I'll update and move back deleteAWSEndpointServices to their original place

**Comment by enxebre** [comment]
last run passed, but hit flake
```
{Failed  === RUN   TestCreateClusterCustomConfig/EnsureNetworkPolicies/EnsureComponentsHaveNeedManagementKASAccessLabel
    util.go:796: 
        Unexpected error:
            <*errors.errorString | 0xc003e7d6e0>: 
            not all expected components have NeedManagementKASAccessLabel:   []string(Inverse(Sort, []string{
              	... // 10 identical elements
              	"ignition-server",
              	"machine-approver",
            + 	"olm-collect-profiles-28185803-xzl2b",
              	"private-router",
              }))
```
need https://github.com/openshift/hypershift/pull/2874

**Pull request merged**